### PR TITLE
ethdb: Partitioned storage

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request implements a new `ethdb.DB` which stores separate tables for different data and some of those tables are partitioned by block number. The code is successfully running and syncing against `testnet` but the test suite needs to be fixed.

@jmank88 Is it possible to spin up a single instance to continuously run against testnet using this branch? We're going to want to run it for a while before feeling comfortable swapping all the nodes for the new storage system.

## TODO

- [x] Test suite
- [x] Migration

### CODE REVIEW

File: internal/debug/flags.go [hunk 1]
```diff
@@ -127,6 +127,9 @@ func Setup(ctx *cli.Context) error {
 
 	// pprof server
 	if ctx.GlobalBool(pprofFlag.Name) {
+		runtime.SetBlockProfileRate(1)
+		runtime.SetMutexProfileFraction(1)
```
Comment (jmank88) [inline internal/debug/flags.go]: Should we wire these up to flags?

File: core/blockchain.go [hunk 2]
```diff
@@ -1486,12 +1517,8 @@ func (bc *BlockChain) reorg(ctx context.Context, oldBlock, newBlock *types.Block
 	}
 	// When transactions get deleted from the database that means the
 	// receipts that were created in the fork must also be deleted
-	batch := bc.db.NewBatch()
 	for _, tx := range diff {
-		DeleteTxLookupEntry(batch, tx.Hash())
-	}
-	if err := batch.Write(); err != nil {
-		return err
+		DeleteTxLookupEntry(bc.db.GlobalTable(), tx.Hash())
```
Comment (jmank88) [inline core/blockchain.go]: Can this batch be preserved?

File: core/blockchain.go [hunk 3]
```diff
@@ -943,7 +965,12 @@ func (bc *BlockChain) WriteBlockWithState(ctx context.Context, block *types.Bloc
 	if err := bc.hc.WriteTd(block.Hash(), block.NumberU64(), externTd); err != nil {
 		return NonStatTy, err
 	}
-	if err := WriteBlock(bc.db, block); err != nil {
+	// Write other block data using a batch.
+	gbatch := bc.db.GlobalTable().NewBatch()
+	bbatch := bc.db.BodyTable().NewBatch()
+	hbatch := bc.db.HeaderTable().NewBatch()
+	rbatch := bc.db.ReceiptTable().NewBatch()
```
Comment (jmank88) [inline core/blockchain.go]: This batch section was updated to start batching down below at `WriteBlockReceipts` - I had pulled it in wrong from upstream - but I'm not sure it makes a big difference either way. Maybe it depends on whether we still want to write the block even if something else goes wrong a little later.

File: eth/backend.go [hunk 4]
```diff
@@ -159,38 +145,40 @@ func New(sctx *node.ServiceContext, config *Config) (*GoChain, error) {
...
+				switch prefix {
+				case 'h':
+					return eth.blockchain.CurrentHeader().Number.Uint64()
+				case 'b', 'r':
+					return eth.blockchain.CurrentBlock().Number().Uint64()
+				}
+				return 0
+			})
+		}
+	*/
```
Comment (jmank88) [inline eth/backend.go]: Nuke it?

File: eth/db_upgrade.go [hunk 5]
```diff
@@ -137,3 +138,4 @@ func upgradeDeduplicateData(db ethdb.Database) func() error {
 		return <-errc
 	}
 }
+*/
```
Comment (jmank88) [inline eth/db_upgrade.go]: Nuke it? IIRC this goes away soon anyways

File: eth/gen_config.go [hunk 6]
```diff
@@ -37,8 +36,8 @@ func (c Config) MarshalTOML() (interface{}, error) {
 		TxPool                  core.TxPoolConfig
 		GPO                     gasprice.Config
 		EnablePreimageRecording bool
-		DocRoot                 string         `toml:"-"`
-		Archive                 archive.Config `toml:",omitempty"`
+		DocRoot                 string `toml:"-"`
+		// Archive                 archive.Config `toml:",omitempty"`
```
Comment (jmank88) [inline eth/gen_config.go]: `rm` all the commented archive stuff in here?

File: ethdb/database.go [hunk 7]
```diff
@@ -327,70 +328,67 @@ func (b *ldbBatch) Reset() {
...
-func (tb *tableBatch) ValueSize() int {
+func (tb *ldbTableBatch) ValueSize() int {
 	return tb.batch.ValueSize()
 }
 
-func (tb *tableBatch) Reset() {
+func (tb *ldbTableBatch) Reset() {
 	tb.batch.Reset()
 }
+*/
```
Comment (jmank88) [inline ethdb/database.go]: delete?

File: ethdb/database_test.go [hunk 8]
```diff
@@ -212,3 +213,4 @@ func testParallelPutGet(db ethdb.Database, t *testing.T) {
 	}
 	pending.Wait()
 }
+*/
```
Comment (jmank88) [inline ethdb/database_test.go]: delete?

File: ethdb/s3/s3.go [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+		if os.IsNotExist(err) {
+			// nop
+		} else if err != nil {
+			return nil, err
+		} else if !fi.IsDir() {
+			if err := os.Remove(s.Path()); err != nil {
+				return nil, err
+			}
+		}
+	*/
```
Comment (jmank88) [inline ethdb/s3/s3.go]: ?

File: internal/debug/flags.go (hunk 1 above)
Comment (jmank88) [reply internal/debug/flags.go]: ^ bump

File: core/blockchain.go [hunk 10]
```diff
@@ -140,7 +140,7 @@ type BlockChain struct {
 // NewBlockChain returns a fully initialised block chain using information
 // available in the database. It initialises the default Ethereum Validator and
 // Processor.
-func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *params.ChainConfig, engine consensus.Engine, vmConfig vm.Config) (*BlockChain, error) {
+func NewBlockChain(ctx context.Context, db common.Database, cacheConfig *CacheConfig, chainConfig *params.ChainConfig, engine consensus.Engine, vmConfig vm.Config) (*BlockChain, error) {
```
Comment (jmank88) [inline core/blockchain.go]: This ctx was dropped, but nbd to clean up later (there are a lot of callers...).

**Comment by benbjohnson** [comment]
@jmank88 Items fixed. Ready for review again.

**Pull request merged**